### PR TITLE
Blog Api Backend

### PR DESCRIPTION
**Pull request by coder1033**
### Endpoints
There are a total of 15 endpoints -

- **Public**:
    ```
    Get all posts -
      URI: GET /read-post
      Status: 200
      Response: {
                  posts: [
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                 ]
               }
    ```
    ```
    Get single post -
      URI: GET /read-post/:id
      Status: 200
      Response: {
                  post:
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                }
    ```
- **Auth**:
    ```
    User Sign Up -
      URI: POST /auth/sign-up
      Status: 200
      Request: {
                    firstName,
                    lastName,
                    email,
                    password
                } 
    ```
    ```
    User Log In -
      URI: POST /auth/log-in
      Status: 200
      Request: {
                    email,
                    password
                }
      Response Header: {
                          user-auth-token: <jwt token>
                        }

    ```
    ```
    Admin Sign Up -
      URI: POST /auth/admin/sign-up
      Status: 200
      Request: {
                    firstName,
                    lastName,
                    email,
                    password,
                    adminKey
                }
    ```
    ```
    Admin Log In -
      URI: POST /auth/admin/log-in
      Status: 200
      Request: {
                    email,
                    password
                }
      Response Header: {
                          admin-auth-token: <jwt token>
                        }

    ```

- **User**:
    ```
    Get all posts -
      URI: GET /user/read-post
      Status: 200
      Request Header: {
                        user-auth-token: <jwt token>
                      }
      Response: {
                  posts: [
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                 ]
               }
    ```
    ```
    Get single post -
      URI: GET /user/read-post/:id
      Status: 200
      Request Header: {
                        user-auth-token: <jwt token>
                      }
      Response: {
                  post:
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                }
    ```
    ```
    Like post -
      URI: POST /admin/like-post
      Status: 200
      Request: {
                  postId,
                  like: <like count>
                }
      Request Header: {
                        user-auth-token: <jwt token>
                      }
    ```
- **Admin**:
    ```
    Get all posts -
      URI: GET /admin/read-post
      Status: 200
      Request Header: {
                        amdin-auth-token: <jwt token>
                      }
      Response: {
                  posts: [
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                 ]
               }
    ```
    ```
    Get single post -
      URI: GET /admin/read-post/:id
      Status: 200
      Request Header: {
                        admin-auth-token: <jwt token>
                      }
      Response: {
                  post:
                      {
                          _id,
                          like,
                          admin,
                          title,
                          content,
                          createdAt,
                          updatedAt
                      }
                }
    ```
    ```
    Create post - 
      URI: POST /admin/create-post
      Status: 200
      Request: {
                    adminId,
                    title,
                    content
                }
      Request Header: {
                          admin-auth-token: <jwt token>
                      }

    ```
    ```
    Update post -
      URI: POST /admin/update-post
      Status: 200
      Request: {
                    adminId,
                    postId,
                    title,
                    content
                }
      Request Header: {
                          admin-auth-token: <jwt token>
                      }
    ```
    ```
    Delete post -
      URI: POST /admin/delete-post
      Status: 200
      Request: {
                    postId
                }
      Request Header: {
                          admin-auth-token: <jwt token>
                      }
    ```
    ```
    Like post -
      URI: POST /admin/like-post
      Status: 200
      Request: {
                  postId,
                  like: <like count>
                }
      Request Header: {
                        user-auth-token: <jwt token>
                      }
    ```

### Database schema

**Admin**
```
    firstName: String
    lastName: String
    email: String
    password: String
    createdAt: Date
    updatedAt: Date
```

**User**
```
    firstName: String
    lastName: String
    email: String
    password: String
    createdAt: Date
    updatedAt: Date
```

**Post**
```
    title: String
    content: String
    admin: Admin
    like: Number
    createdAt: Date
    updatedAt: Date
```

### Tests

**Automated test cases added**
- None

**Manual test cases run**
- Checks all the endpoints
- Creates few test data

### CODE REVIEW

File: blog-api-backend/app.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// npm imports
+const cookieParser = require('cookie-parser');
+const cors = require("cors");
+const createError = require('http-errors');
```
Comment (balwant-jtc) [inline blog-api-backend/app.js:4]: Be consistent with quotes use single or double quotes but don't use both.

File: blog-api-backend/app.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+const verifyToken = require("./middlewares/verify-token");
+
+// environment variables
+dotenv.config();
+
+// database connection
+const mongoDb = process.env.DB_CONNECTION_STRING;
+mongoose.connect(mongoDb, { useUnifiedTopology: true, useNewUrlParser: true });
+const db = mongoose.connection;
+db.on("error", console.error.bind(console, "mongo connection error"));
```
Comment (balwant-jtc) [inline blog-api-backend/app.js:25]: use logger instead of console

File: blog-api-backend/controllers/admin-controller.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+// delete post
+exports.deletePost = async function (req, res, next) {
+  const { postId } = req.body;
+  const isValidPostId = mongoose.Types.ObjectId.isValid(req.body.postId);
+
+  if (!isValidPostId) {
+    return res.status(400).send({ error: "Invalid post Id." });
+  } else {
+    try {
+      const post = await Post.findById(postId);
```
Comment (balwant-jtc) [inline blog-api-backend/controllers/admin-controller.js:160]: No need to check for the id just perform delete operation.

File: blog-api-backend/controllers/admin-controller.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+      return next(err);
+    }
+  }
+
+  if (!isValidPostId) {
+    errors.push(invalidPostId);
+  } else {
+    post._id = postId;
+    try {
+      const post = await Post.findById(postId);
```
Comment (balwant-jtc) [inline blog-api-backend/controllers/admin-controller.js:129]: No need to check for the id just issue the update command and check for the response if it returns the object or not, and on that basis return appropriate status code.

File: blog-api-backend/controllers/admin-controller.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    msg: "Invalid post Id.",
+    param: "postId",
+    location: "body",
+  };
+
+  if (!isValidPostId) {
+    errors.push(error);
+  } else {
+    try {
+      const post = Post.findById(postId);
```
Comment (balwant-jtc) [inline blog-api-backend/controllers/admin-controller.js:195]: Same here, don't check for postId explicitly

File: blog-api-backend/app.js (hunk 2 above)
Comment (coder1033) [reply blog-api-backend/app.js:25]: It's a boilerplate code generated by express. So should I update it?

File: blog-api-backend/app.js (hunk 1 above)
Comment (coder1033) [reply blog-api-backend/app.js:4]: Okay

File: blog-api-backend/controllers/auth-controller.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    const compare = await bcrypt.compare(password, user.password);
+    if (!compare) {
+      return res.status(400).send({ user: { email, password }, error });
+    }
+
+    // create and asign a token
+    const token = jwt.sign({ _id: user.id }, process.env.USER_TOKEN_SECRET, {
+      expiresIn: "30d",
```
Comment (balwant-jtc) [inline blog-api-backend/controllers/auth-controller.js:81]: Ideally configurable things like token expiryTime should be in some separate constants file.

File: blog-api-backend/controllers/auth-controller.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    lastName,
+    email,
+    password,
+  });
+
+  const errors = validationResult(req).array();
+
+  try {
+    const found = await Admin.findOne({ email });
+    console.log(found);
```
Comment (balwant-jtc) [inline blog-api-backend/controllers/auth-controller.js:104]: remove all console logs and use logger

File: blog-api-backend/models/admin.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require("mongoose");
+
+const Schema = mongoose.Schema;
+
+const AdminSchema = new Schema(
+  {
+    firstName: { type: String, required: true, minlength: 3, maxlength: 50 },
+    lastName: { type: String, required: true, minlength: 3, maxlength: 50 },
+    email: { type: String, required: true },
```
Comment (balwant-jtc) [inline blog-api-backend/models/admin.js:9]: Should have unique index over email column, read about database indexes in free time.

File: blog-api-backend/routes/admin.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  adminController.createPost
+);
+
+router.put(
+  "/update-post",
+  adminValidator.createPost,
+  adminController.updatePost
+);
+
+router.delete("/delete-post", adminController.deletePost);
```
Comment (balwant-jtc) [inline blog-api-backend/routes/admin.js:22]: rest verb is self explanatory about the action here, so no need to be specific with routes, just use /posts/:id with delete verb

File: blog-api-backend/routes/admin.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+router.put("/like-post", adminValidator.like, adminController.likePost);
+
+router.get("/read-post", adminController.readAllPost);
```
Comment (balwant-jtc) [inline blog-api-backend/routes/admin.js:26]: rest verb is self explanatory about the action here, so no need to be specific with routes, just use /posts with get verb

File: blog-api-backend/routes/index.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+const express = require("express");
+
+// npm imports
+const userController = require("../controllers/user-controller");
+
+const router = express.Router();
+
+router.use("/read-post/:id", userController.readSinglePost);
+
+router.use("/read-post", userController.readAllPost);
```
Comment (balwant-jtc) [inline blog-api-backend/routes/index.js:11]: use /posts/:id  -> for single post
/posts -> for all posts

File: blog-api-backend/routes/user.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+// npm imports
+const express = require("express");
+
+// local imports
+const userController = require("../controllers/user-controller");
+
+const router = express.Router();
+
+router.get("/read-post/:id", userController.readSinglePost);
```
Comment (balwant-jtc) [inline blog-api-backend/routes/user.js:9]: Same as previous comment

**Review by balwant-jtc — APPROVED** [review]
approved

**Pull request merged**